Fix: Improve LowCardinality validation for model contracts

jorgeparavicini opened the pull request:
## Summary
This pull request resolves #303 by allowing the definition of columns as `LowCardinality` in the contracts. Before, defining a column of type `String` to be of type `LowCardinality(String)` caused an error if the constraints were enforced; otherwise, the `LowCardinality` was just ignored. 

Additionally, this PR allows defining special types that ignore the basic string comparison in the column constraints. 
At the moment, only the `SimpleAggregateFunction` is marked as a special type

## Checklist
Delete items not relevant to your PR:
- [x] Unit and integration tests covering the common scenarios were added
- [x] A human-readable description of the changes was provided to include in CHANGELOG

koletzilla commented on dbt/include/clickhouse/macros/column_spec_ddl.sql:
```diff
@@ -30,10 +30,31 @@
+      {%- endif -%}
+    {%- endfor -%}
+
+    {% if not ns.is_special_type %}
+      {%- if yaml_data_type.startswith('LowCardinality') -%}
+        {#-- If contract is LowCardinality, extract the inner type for comparison --#}
+        {%- set yaml_data_type = yaml_data_type[15:-1] -%}
+      {%- endif -%}
+
+      {%- if sql_data_type != yaml_data_type -%}
+        {#-- Column data types don't match #}
+        {%- do exceptions.raise_contract_error(yaml_columns, sql_columns) -%}
+      {%- endif -%}
+    {% endif %} 
```
the `{%- if sql_data_type != yaml_data_type -%}` is inside the `not ns.is_special_type`, so if the data type is "is_special_type", it don't gets validated.

As we are introducing some easy validations also for SimpleAggregateFunction here, what do you think if we also remove the `SimpleAggregateFunction` and get the inside type validated? I have modify your changes to this so:
- if `SimpleAggregateFunction`, it gets the second argument.
- same logic with `LowCardinality`
- then compare `sql_data_type != ns.yaml_data_type`

```jinja2
    {%- set ns = namespace(yaml_data_type=yaml_col[0]['data_type']) -%}
    {%- set sql_data_type = sql_col['data_type'] -%}

    {%- set special_types = ['SimpleAggregateFunction'] -%}
    {%- for special_type in special_types -%}
      {%- if ns.yaml_data_type.startswith(special_type) -%}
        {%- set ns.yaml_data_type = ns.yaml_data_type.split('(', 1)[1].split(',')[1].strip()[:-1] -%}
        {%- break -%}
      {%- endif -%}
    {%- endfor -%}

    {%- if ns.yaml_data_type.startswith('LowCardinality') -%}
      {#-- If contract is LowCardinality, extract the inner type for comparison --#}
      {%- set ns.yaml_data_type = ns.yaml_data_type[15:-1] -%}
    {%- endif -%}

    {%- if sql_data_type != ns.yaml_data_type -%}
      {#-- Column data types don't match #}
      {%- do exceptions.raise_contract_error(yaml_columns, sql_columns) -%}
    {%- endif -%}
```

jorgeparavicini commented:
Hi @koletzilla, great points, I've updated the implementation accordingly. I also modified the macro so that we have a single location where we need to add new "special types". I added both the LowCardinality and the SimpleAggregateFunction there now.
